accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,671 @@
+
+Internally in the runtime, the "CallKind" information will no longer be tracked by the function pointer type. Currently it is tracked, but that will change for unmodified types to just detecting managed (via "default" and "vararg") to avoid inconsistent semantics across the older "CallKind" and newer "modopts" encoding formats.
+
+### _Off-topic thoughts on using custom modifiers to represent calling conventions_
+Using the CallConv* classes to represent calling conventions was added in V6 because new callling conventions needed to be added and would not fit within the 4 bits that were being used to encode the existing calling conventions based on ECMA-335. So the decision, arguably not the best, was made to encode new calling conventions using custom modifiers since that would not affect existing metadata readers or require a ECMA-335 change. However, there are alternative approaches which may or may not have been considered -- one such approach would be to change the encompassing byte to be a 32-bit compressed integer encoding instead which .NET/CLI uses extensively elsewhere. This compressed integer approach is possible since the encompassing byte did not use bit 7 (`0x80`) which is the flag for a compressed integer to switch from one byte to two bytes. This would require a ECMA-335 change which would break other metadata readers once two+ bytes are used.
+
+## Modified types
+Obtaining all custom modifiers including `ref\out\in` is useful for those who read metadata for varying reasons. The proposal to support that is to expose functionality that returns a "modified type" which, for function pointers, is a wrapper over the unmodified function pointer and intercepts calls to `GetFunctionPointerParameters()` to return a "modified type" for each parameter. From that parameter modified type, `GetRequiredCustomModifiers()` and `GetOptionalCustomModifiers()` can be used.
+
+A modified type always start with a "root" function pointer obtained from `FieldInfo.GetModifiedFieldType()`, `PropertyInfo.GetModifiedPropertyType()` and `ParameterInfo.GetModifiedParameterType()`. **discussion topic**: a modified type "tree" ends with the modified parameter types. For example, if a field is a function pointer that has parameters and those parameters are also function pointers, the secondary function pointers are not accessible as modified types. This limitation was added to simplify the design. Without it, an N-level tree of modified types would need to be created, recusively. It is possible to add this in the future, however.
```

Comment:
> This limitation was added to simplify the design.

Does it simplify the design, or just the implementation? I think it is just the latter, and it is pretty marginal simplification.

Reply:
> It is possible to add this in the future, however.

Is it going to be a breaking change to add it later?

Reply:
Both the design and implementation. If we don't support "nested" types, the implementation becomes trivial -- implemented in managed code only, using the existing "signature" field already there for FieldInfo, PropertyInfo and MethodInfo.

If we supported "nested" types, then the [discussion here](https://github.com/dotnet/designs/pull/282#discussion_r1049031775) becomes relevant. Do we want to support obtaining custom mods from all of the following function pointers?
```
// Assumed
delegate*<int> s_0 = null;

// Arrays?
delegate*<int>[] a_0 = new delegate*<int>[] { };
delegate*<int>[][] a_1 = null!;
delegate*<int>[,] a_2 = null!;
// etc.

// Pointers?
delegate*<int>* p_0 = null;
delegate*<int>** p_1 = null;
delegate*<int>*** p_2 = null;
// etc.

// Arrays + pointers?
delegate*<int>*[] ap_0 = null!;
delegate*<int>**[] ap_1 = null!;
delegate*<int>*[][] ap_2 = null!;
// etc.

// Nested fnptrs?
delegate*<delegate*<int>> fp_0 = null;
// with pointers
delegate*<delegate*<int>*> fp_1 = null;
// with arrays
delegate*<delegate*<int>[]> f2_3 = null;
// with arrays + pointers
delegate*<delegate*<int>*[]> f2_5 = null;
// etc...
```

Reply:
I think we should try to support the above, but need to understand how these nested\inner types are populated -- eager vs. lazy for example. If eager, and in the same "loop", then this should be trivial to support.

Reply:
It should be fine to start with eager population for simplicity. (We can always switch to lazy if the eager population of the full modified type shows up as a performance issue.)

Reply:
Resolving; the current PR supports the above in a lazy-load manner. It also supports generics such as `Tuple<delegate*<void>[]` which is allowed by the compiler since the function pointer is not directly the generic signature type.